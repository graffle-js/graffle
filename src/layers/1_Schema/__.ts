export * as SchemaKit from './_.js'
export * from './_.js'
