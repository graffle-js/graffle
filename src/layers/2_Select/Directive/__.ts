export * as Directive from './_.js'
