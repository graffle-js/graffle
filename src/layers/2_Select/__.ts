export * as Select from './_.js'
