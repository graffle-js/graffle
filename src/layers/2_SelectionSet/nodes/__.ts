export * as Nodes from './_.js'
