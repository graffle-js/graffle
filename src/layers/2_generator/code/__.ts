import { createCodeGenerator } from '../createCodeGenerator.js'
import { moduleName_ } from './_.js'

export const { generate: generate__, moduleName: moduleName__ } = createCodeGenerator(
  `__`,
  (_config) => {
    const code: string[] = []
    // todo allow user to customize namespace. Should "default" map to "Graffle" and then anything else is explicitly passed through?
    code.push(
      `export * as Graffle from './${moduleName_}.js'`,
    )

    return code.join(`\n\n`)
  },
)
