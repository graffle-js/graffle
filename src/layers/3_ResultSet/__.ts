export * as ResultSet from './infer.js'
