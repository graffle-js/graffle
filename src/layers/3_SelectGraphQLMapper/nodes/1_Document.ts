import type { SchemaDrivenDataMap, SchemaKit } from '../../../entrypoints/utilities-for-generated.js'
import type { Grafaid } from '../../../lib/grafaid/__.js'
import { Nodes } from '../../../lib/grafaid/_Nodes.js'
import type { Select } from '../../2_Select/__.js'
import { toGraphQLOperationDefinition } from './2_OperationDefinition.js'

const defaultOperationName = `$default`

export const toGraphQLDocument = (
  graffleDocument: Select.Document.DocumentNormalized,
  options?: Options,
): Encoded => {
  const operationsAndVariables = Object
    .values(graffleDocument.operations)
    .map(graffleOperation => {
      const sddm = options?.sddm?.roots[graffleOperation.rootType]
      return toGraphQLOperationDefinition(sddm, graffleOperation, options)
    })

  const graphqlDocument = Nodes.Document({
    definitions: operationsAndVariables.map(_ => _.operation),
  })

  const operationsVariables = Object.fromEntries(operationsAndVariables.map((_): [string, Grafaid.Variables] => {
    const name = _.operation.name?.value ?? defaultOperationName
    return [name, _.variables]
  }))

  return {
    document: graphqlDocument,
    operationsVariables,
  }
}

export interface Options {
  sddm?: SchemaDrivenDataMap | null
  scalars?: SchemaKit.Hybrid.Scalar.ScalarMap
  operationVariables?: boolean
}

export interface Encoded {
  document: Grafaid.Document.DocumentNode
  operationsVariables: Record<string, Grafaid.Variables>
}
