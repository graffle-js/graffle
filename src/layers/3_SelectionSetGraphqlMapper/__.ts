export * as SelectionSetGraphqlMapper from './_.js'
