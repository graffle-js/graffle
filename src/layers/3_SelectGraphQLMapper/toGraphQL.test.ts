import { expect, test } from 'vitest'
import { db } from '../../../tests/_/schemas/db.js'
import { schemaDrivenDataMap } from '../../../tests/_/schemas/kitchen-sink/graffle/modules/SchemaDrivenDataMap.js'
import type * as SelectionSets from '../../../tests/_/schemas/kitchen-sink/graffle/modules/SelectionSets.js'
import { Grafaid } from '../../lib/grafaid/__.js'
import { Select } from '../2_Select/__.js'
import { toGraphQLDocument } from './nodes/1_Document.js'

type CasesDescriptiveQuery = [
  description: string,
  selectionSet: SelectionSets.Query,
  options?: { operationName?: string },
]
const testEachQueryWithDescription = test.for.bind(test)<CasesDescriptiveQuery>

const tester = (input: { variables: boolean }) =>
  [
    `( variables: ${String(input.variables)} ) - Query - %s`,
    (args: CasesDescriptiveQuery) => {
      const [description, graffleQuery, options] = args

      const { document, operationsVariables } = toGraphQLDocument(
        Select.Document.createDocumentNormalizedFromQuerySelection(
          graffleQuery as any,
          options?.operationName,
        ),
        {
          sddm: schemaDrivenDataMap,
          operationVariables: input.variables,
        },
      )

      const beforeAndAfter = `\n`
        + `\n--------------GRAFFLE QUERY-------------\n`
        + JSON.stringify(graffleQuery, null, 2)
        + `\n--------GRAPHQL DOCUMENT & VARIABLES--------\n`
        + Grafaid.Document.print(document)
        + `\n----------------\n`
        + JSON.stringify(operationsVariables, null, 2)
        + `\n`

      expect(beforeAndAfter).toMatchSnapshot(description)
    },
  ] as const

// todo test a case where we provide an operation name
// dprint-ignore
const cases = testEachQueryWithDescription([
  [`fg - one`             , { ___: { __typename: true } }],
  [`fg - multiple`        , { ___: [{ __typename: true }, { abcEnum: true }] }],
  [`fg - interface`       , { interface: { ___: { __typename: true } } }],
  [`fg - in union`        , { unionFooBar: { ___: { __typename: true } } }],
  [`fg - directive`       , { ___: { __typename: true, $include: true } }],
  // union
  [`union - __typename (no fragment)`  , { unionFooBar: { __typename: true } }],
  [`union - scalar`                    , { unionFooBar: { ___on_Bar: { int: true } } }],
  [`union - directive`                 , { unionFooBar: { ___on_Bar: { $skip: true, int: true } } }],
  // s({ unionFooBar: { __on_Bar: {} } }), // todo should be static type error
  // alias
  [`alias - scalar`                    , { id: [`x`, true] }],
  [`alias - scalar x2`                 , { id: [[`x`, true], [`id2`, true]] }],
  [`alias - scalar directive`          , { id: [`x`, { $skip: true }] }],
  [`alias - scalar directive+select`   , { object: [`x`, { $skip: true, id: true }] }],
    // arguments
  [`args - on union`                             , { result: { $: { $case: `Object1` }, __typename: true } }],
  [`args - string with args`                     , { stringWithArgs: { $: { boolean: true, float: 1 } } }],
  [`args - alias`                                , { stringWithArgs: [[`a`, { $: { id: `` }}]] }],
  [`args - x2 same`                              , { stringWithArgs: [[`a`, { $: { id: `` }}], [`b`,{$:{id:``}}]] }],
  [`args - string with args (empty object)`      , { stringWithArgs: { $: {} } }],
  // arguments custom scalars
  [`args - custom scalar - arg field`                                , { dateArg: { $: { date: db.date0 } } }],
  [`args - custom scalar - arg field in non-null`                    , { dateArgNonNull: { $: { date: db.date0 } } }], 
  [`args - custom scalar - arg field in list`                        , { dateArgList: { $: { date: [db.date0, new Date(1)] } } }], 
  [`args - custom scalar - arg field in list (null)`                 , { dateArgList: { $: { date: null } } }],
  [`args - custom scalar - arg field in non-null list (with list)`   , { dateArgNonNullList: { $: { date: [db.date0, new Date(1)] } } }],
  [`args - custom scalar - arg field in non-null list (with null)`   , { dateArgNonNullList: { $: { date: [null, db.date0] } } }],
  [`args - custom scalar - arg field in non-null list non-null`      , { dateArgNonNullListNonNull: { $: { date: [db.date0, new Date(1)] } } }],
  [`args - custom scalar - input object field`                       , { dateArgInputObject: { $: { input: { idRequired: ``, dateRequired: db.date0, date: new Date(1) } } } }],
  [`args - custom scalar - nested input object field`                , { InputObjectNested: { $: { input: { InputObject: { idRequired: ``, dateRequired: db.date0, date: new Date(1) } } } } }],
  // s({ objectWithArgs: { $: { id: `` } } }), // todo should be static error
  // s({ objectWithArgs: { $: {} } }), // todo should be static error
  [`args - object with args`, { objectWithArgs: { $: { id: `` }, id: true } }],
  [`args - object with args (empty object)`      , { objectWithArgs: { $: {}, id: true } }],
  // $include
  [`$include`                             , { object: { $include: true, id: true } }],
  [`$include (false)`                     , { object: { $include: false, id: true } }],
  [`$include (undefined)`                 , { object: { $include: undefined, id: true } }],
  [`$include (if true)`                   , { object: { $include: { if: true }, id: true } }],
  [`$include (if false)`                  , { object: { $include: { if: false }, id: true } }],
  [`$include (if undefined)`              , { object: { $include: { if: undefined }, id: true } }],
  [`$include (empty object)`              , { object: { $include: {}, id: true } }],
  // skip
  [`$skip (true)`                         , { object: { $skip: true, id: true } }],
  [`$skip (false)`                        , { object: { $skip: false, id: true } }],
  [`$skip (undefined)`                    , { object: { $skip: undefined, id: true } }],
  [`$skip (if true)`                      , { object: { $skip: { if: true }, id: true } }],
  [`$skip (if false)`                     , { object: { $skip: { if: false }, id: true } }],
  [`$skip (if undefined)`                 , { object: { $skip: { if: undefined }, id: true } }],
  [`$skip (empty object)`                 , { object: { $skip: {}, id: true } }],
  [`object nested scalar $skip`           , { objectNested: { object: { string: true, id: true, int: { $skip: true } } } }],
  // other
  [`custom operation name`                , { id: true }, {operationName: `foobar` }],
  [`other`                                , { __typename: true }],
  [`string`                               , { string: true }],
  // [{ string: 1 }],
  // s({ string: false }), // todo should be static error
  [`id string false`                      , { id: true, string: false }],
  // [{ id: true, string: 0 }],
  [`id string undefined`                  , { id: true, string: undefined }],
  [`object scalar`                        , { object: { id: true } }],
  [`object nested`                        , { objectNested: { object: { string: true, id: true, int: false } } }],
])
cases(...tester({ variables: true }))
cases(...tester({ variables: false }))
