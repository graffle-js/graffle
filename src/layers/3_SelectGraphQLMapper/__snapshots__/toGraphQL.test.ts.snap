// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`alias > Query - { id: [ 'x', { 'undefined': true } ] } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "id": [
    "x",
    {
      "$skip": true
    }
  ]
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  x: id @skip(if: true)
}
----------------
null
"
`;

exports[`alias > Query - { id: [ 'x', true ] } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "id": [
    "x",
    true
  ]
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  x: id
}
----------------
null
"
`;

exports[`alias > Query - { id: [ [ 'x', true ], [ 'id2', true ] ] } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "id": [
    [
      "x",
      true
    ],
    [
      "id2",
      true
    ]
  ]
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  x: id
  id2: id
}
----------------
null
"
`;

exports[`alias > Query - { object: [ 'x', { 'undefined': true, id: true } ] } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": [
    "x",
    {
      "$skip": true,
      "id": true
    }
  ]
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  x: object @skip(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`arguments > Query - { objectWithArgs: { '$': { id: '' }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "objectWithArgs": {
    "$": {
      "id": ""
    },
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  objectWithArgs(id: "") {
    id
  }
}
----------------
null
"
`;

exports[`arguments > Query - { objectWithArgs: { '$': {}, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "objectWithArgs": {
    "$": {},
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  objectWithArgs {
    id
  }
}
----------------
null
"
`;

exports[`arguments > Query - { result: { '$': { 'undefined': 'Object1' }, __typename: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "result": {
    "$": {
      "$case": "Object1"
    },
    "__typename": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  result(case: Object1) {
    __typename
  }
}
----------------
null
"
`;

exports[`arguments > Query - { stringWithArgs: { '$': { boolean: true, float: 1 } } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "stringWithArgs": {
    "$": {
      "boolean": true,
      "float": 1
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  stringWithArgs(boolean: true, float: 1)
}
----------------
null
"
`;

exports[`arguments > Query - { stringWithArgs: { '$': {} } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "stringWithArgs": {
    "$": {}
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  stringWithArgs
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field > arg field 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArg": {
    "$": {
      "date": "1970-01-01T00:00:00.000Z"
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArg(date: {})
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in list (null) > arg field in list (null) 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgList": {
    "$": {
      "date": null
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgList(date: null)
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in list > arg field in list 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgList": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgList(date: [{}, {}])
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in non-null > arg field in non-null 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgNonNull": {
    "$": {
      "date": "1970-01-01T00:00:00.000Z"
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgNonNull(date: {})
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in non-null list (with list) > arg field in non-null list (with list) 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgNonNullList": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgNonNullList(date: [{}, {}])
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in non-null list (with null) > arg field in non-null list (with null) 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgNonNullList": {
    "$": {
      "date": [
        null,
        "1970-01-01T00:00:00.000Z"
      ]
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgNonNullList(date: [null, {}])
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - arg field in non-null list non-null > arg field in non-null list non-null 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgNonNullListNonNull": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgNonNullListNonNull(date: [{}, {}])
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - input object field > input object field 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "dateArgInputObject": {
    "$": {
      "input": {
        "idRequired": "",
        "dateRequired": "1970-01-01T00:00:00.000Z",
        "date": "1970-01-01T00:00:00.001Z"
      }
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  dateArgInputObject(input: {idRequired: "", dateRequired: {}, date: {}})
}
----------------
null
"
`;

exports[`arguments > custom scalars > Query - nested input object field > nested input object field 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "InputObjectNested": {
    "$": {
      "input": {
        "InputObject": {
          "idRequired": "",
          "dateRequired": "1970-01-01T00:00:00.000Z",
          "date": "1970-01-01T00:00:00.001Z"
        }
      }
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  InputObjectNested(
    input: {InputObject: {idRequired: "", dateRequired: {}, date: {}}}
  )
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': { if: false }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": {
      "if": false
    },
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: false) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': { if: true }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": {
      "if": true
    },
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': { if: undefined }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': {}, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': false, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": false,
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: false) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': true, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$include": true,
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @include(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $include > Query - { object: { 'undefined': undefined, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': { if: false }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": {
      "if": false
    },
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: false) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': { if: true }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": {
      "if": true
    },
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': { if: undefined }, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': {}, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': false, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": false,
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: false) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': true, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "$skip": true,
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object @skip(if: true) {
    id
  }
}
----------------
null
"
`;

exports[`directive > $skip > Query - { object: { 'undefined': undefined, id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object {
    id
  }
}
----------------
null
"
`;

exports[`inline fragment > field group > Query - directive > directive 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "___": {
    "__typename": true,
    "$include": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  ... @include(if: true) {
    __typename
  }
}
----------------
null
"
`;

exports[`inline fragment > field group > Query - interface > interface 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "interface": {
    "___": {
      "__typename": true
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  interface {
    ... {
      __typename
    }
  }
}
----------------
null
"
`;

exports[`inline fragment > field group > Query - multiple > multiple 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "___": [
    {
      "__typename": true
    },
    {
      "abcEnum": true
    }
  ]
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  ... {
    __typename
  }
  ... {
    abcEnum
  }
}
----------------
null
"
`;

exports[`inline fragment > field group > Query - one > one 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "___": {
    "__typename": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  ... {
    __typename
  }
}
----------------
null
"
`;

exports[`inline fragment > field group > Query - union > union 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "unionFooBar": {
    "___": {
      "__typename": true
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  unionFooBar {
    ... {
      __typename
    }
  }
}
----------------
null
"
`;

exports[`inline fragment > type condition union > Query - __typename (no fragment) > __typename (no fragment) 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "unionFooBar": {
    "__typename": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  unionFooBar {
    __typename
  }
}
----------------
null
"
`;

exports[`inline fragment > type condition union > Query - directive > directive 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "unionFooBar": {
    "___on_Bar": {
      "$skip": true,
      "int": true
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  unionFooBar {
    ... on Bar @skip(if: true) {
      int
    }
  }
}
----------------
null
"
`;

exports[`inline fragment > type condition union > Query - scalar > scalar 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "unionFooBar": {
    "___on_Bar": {
      "int": true
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  unionFooBar {
    ... on Bar {
      int
    }
  }
}
----------------
null
"
`;

exports[`other > Query - { __typename: true } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "__typename": true
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  __typename
}
----------------
null
"
`;

exports[`other > Query - { id: true, string: false } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "id": true,
  "string": false
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  id
}
----------------
null
"
`;

exports[`other > Query - { id: true, string: undefined } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "id": true
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  id
}
----------------
null
"
`;

exports[`other > Query - { object: { id: true } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "object": {
    "id": true
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  object {
    id
  }
}
----------------
null
"
`;

exports[`other > Query - { objectNested: { object: { string: true, id: true, int: { 'undefined': true } } } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "objectNested": {
    "object": {
      "string": true,
      "id": true,
      "int": {
        "$skip": true
      }
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  objectNested {
    object {
      string
      id
      int @skip(if: true)
    }
  }
}
----------------
null
"
`;

exports[`other > Query - { objectNested: { object: { string: true, id: true, int: false } } } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "objectNested": {
    "object": {
      "string": true,
      "id": true,
      "int": false
    }
  }
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  objectNested {
    object {
      string
      id
    }
  }
}
----------------
null
"
`;

exports[`other > Query - { string: true } 1`] = `
"

--------------GRAFFLE QUERY-------------
{
  "string": true
}
--------GRAPHQL DOCUMENT & VARIABLES--------
{
  string
}
----------------
null
"
`;
